In [104]:
import math

def calculate_probability_twice_or_more(T, N, R):

    # Calculate the number of possible choices
    num_choices = math.comb(T, N)
    
    # Calculate the probability of choosing the specific set of points in one try
    p = 1 / num_choices

    # Calculate the probability of 0 occurrences
    prob_0 = (1 - p) ** R

    # Calculate the probability of 1 occurrence
    prob_1 = R * p * (1 - p) ** (R - 1)

    # Calculate the probability of the result occurring twice or more
    prob_twice_or_more = 1 - prob_0 - prob_1

    print(f"There are {num_choices} possible choices when selecting {N} points out of {T} at random.")
    print(f"The probability of choosing the specific set of points in one try is 1/{num_choices}.")
    print(f"Probability of the result occurring twice or more in {R} attempts: {prob_twice_or_more}")

    return prob_twice_or_more

# Example usage:
T = 1000  # Number of possible choices
N = 30   # Number of values selected
R = 10000   # Number of times the process is repeated

result = calculate_probability_twice_or_more(T, N, R)

### BROKEN WITH LARGE VALUES DUE TO FP ERRORS ###


There are 2429608192173745103270389838576750719302222606198631438800 possible choices when selecting 30 points out of 1000 at random.
The probability of choosing the specific set of points in one try is 1/2429608192173745103270389838576750719302222606198631438800.
Probability of the result occurring twice or more in 10000 attempts: -4.1158899744460876e-54


In [109]:
#### USING MPMATH FOR ARBITRARY PRECISION ####


import math
from mpmath import mp

def calculate_probability_twice_or_more(T, N, R):
    mp.dps = 300  # Setting a high precision

    # Calculate the number of possible choices
    num_choices = math.comb(T, N)
    
    # Calculate the probability of choosing the specific set of points in one try
    p = mp.mpf(1) / mp.mpf(num_choices)

    # Calculate the probability of 0 occurrences
    prob_0 = (mp.mpf(1) - p) ** R

    # Calculate the probability of 1 occurrence
    prob_1 = R * p * (mp.mpf(1) - p) ** (R - 1)

    # Calculate the probability of the result occurring twice or more
    prob_twice_or_more = mp.mpf(1) - prob_0 - prob_1

    print(f"There are {num_choices} possible choices when selecting {N} points out of {T} at random.")
    print(f"The probability of choosing the specific set of points in one try is 1/{num_choices}.")
    print(f"Probability of the result occurring twice or more in {R} attempts: {prob_twice_or_more}")

    return prob_twice_or_more

T = 2000     # Number of possible choices
N = 30      # Number of values selected
R = 1000000   # Number of times the process is repeated
calculate_probability_twice_or_more(T, N, R)


There are 3253215754299955884534500580299601673059442134171635682543627213600 possible choices when selecting 30 points out of 2000 at random.
The probability of choosing the specific set of points in one try is 1/3253215754299955884534500580299601673059442134171635682543627213600.
Probability of the result occurring twice or more in 1000000 attempts: 4.72436927570703226422768161210407661558706625769841440561434843537164669524162746219719933945240745912775140235358087872029351841564851945645880152028033818442500404178900572552164991959820722107056881068171735796252494098698638947807733850993953400448891109292968688535981726380880585832256367589345e-122


mpf('4.7243692757070322642276816121040766155870662576984144056143484353716466952416274621971993394524074591277514023535808787202935184156485194564588015202803381844250040417890057255216499195982072210705688106817173579625249409869863894780773385099395340044889110929296868853598172638088058583225636758934471e-122')

Probability of the same set of signal points being chosen twice or more in 1,000,000 attempts: 4.7 x 10^-122